File: openc3-cosmos-httpjsondemo/targets/HTTPJSONDEMO/lib/http_json_server_interface.rb
# encoding: ascii-8bit

# Copyright 2026 OpenC3, Inc.
# All Rights Reserved.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE.md for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/interfaces/interface'
require 'openc3/packets/packet'
require 'openc3/accessors/json_accessor'
require 'webrick'
require 'json'

module OpenC3
  # HTTP server interface that receives inbound JSON POSTs as telemetry.
  #
  # Path format: POST /targetname/packetname
  # The path segments are uppercased and used to identify the telemetry packet.
  # Item names in the packet definition must match the incoming JSON key names.
  # JsonAccessor is set automatically on received packets — no ACCESSOR declaration
  # needed in telemetry definitions.
  #
  # Supported options:
  #   LISTEN_ADDRESS <ip>  — bind address (default 0.0.0.0)
  class HttpJsonServerInterface < Interface
    OK_BODY = '{"status":"ok"}'.freeze
    OK_HEADERS = {
      'Content-Type'   => 'application/json',
      'Content-Length' => OK_BODY.bytesize.to_s,
      'Connection'     => 'close'
    }.freeze

    def initialize(port = 80)
      super()
      @listen_address = '0.0.0.0'
      @port = Integer(port)
      @server = nil
      @server_thread = nil
      @api_key = nil
      @max_queue_depth = 1000
      @allowed_content_types = ['application/json']
      @request_queue = Queue.new
    end

    # Supported Options
    # LISTEN_ADDRESS   - IP address to bind (default 0.0.0.0)
    # API_KEY          - Secret key callers must supply in X-Api-Key header
    # MAX_QUEUE_DEPTH  - Max inbound requests buffered before returning 503 (default 1000)
    # ALLOW_CONTENT_TYPE - Additional Content-Type to accept (default: application/json)
    def set_option(option_name, option_values)
      super(option_name, option_values)
      case option_name.upcase
      when 'LISTEN_ADDRESS'
        @listen_address = option_values[0]
      when 'API_KEY'
        @api_key = option_values[0]
      when 'MAX_QUEUE_DEPTH'
        @max_queue_depth = Integer(option_values[0])
      when 'ALLOW_CONTENT_TYPE'
        @allowed_content_types << option_values[0].downcase
      end
    end

    def connection_string
      "listening on #{@listen_address}:#{@port}"
    end

    def connect
      # Build the valid-packet lookup so the handler can reject unknown paths before 200 OK.
      # Also set JsonAccessor on every defined packet so the CVT singleton already has the
      # right accessor — ensures telemetry.update! and decom cloning both work without
      # requiring ACCESSOR declarations in the telemetry definitions.
      @valid_packets = {}
      @target_names.each do |target_name|
        begin
          System.telemetry.packets(target_name).each do |pkt_name, pkt|
            (@valid_packets[target_name] ||= {})[pkt_name] = true
            pkt.accessor = JsonAccessor.new(pkt)
          end
        rescue => e
          Logger.error("HttpJsonServerInterface: failed to enumerate packets for #{target_name}\n#{e.formatted}")
        end
      end

      @request_queue = SizedQueue.new(@max_queue_depth)
      @server = WEBrick::HTTPServer.new(
        BindAddress: @listen_address,
        Port: @port,
        Logger: WEBrick::Log.new('/dev/null'),
        AccessLog: []
      )

      @server.mount_proc '/' do |req, res|
        # 1. Authentication
        if @api_key && req['x-api-key'] != @api_key
          res.status = 401
          res['Content-Type'] = 'application/json'
          res.body = JSON.generate({ status: 'error', message: 'Unauthorized' })
          next
        end

        # 2. Path — expect /targetname/packetname
        parts = req.path.sub(%r{^/+}, '').sub(%r{/+$}, '').split('/')
        unless parts.length == 2
          res.status = 400
          res['Content-Type'] = 'application/json'
          res.body = JSON.generate({ status: 'error', message: "Path must be /target/packet, got: #{req.path}" })
          next
        end

        target_name = parts[0].upcase
        packet_name = parts[1].upcase

        # 3. Known packet validation
        unless @valid_packets.dig(target_name, packet_name)
          res.status = 404
          res['Content-Type'] = 'application/json'
          res.body = JSON.generate({ status: 'error', message: "Unknown packet: #{target_name}/#{packet_name}" })
          next
        end

        # 4. Content-Type enforcement
        content_type = req['content-type'].to_s.split(';').first.to_s.strip.downcase
        unless @allowed_content_types.include?(content_type)
          res.status = 415
          res['Content-Type'] = 'application/json'
          res.body = JSON.generate({
            status: 'error',
            message: "Unsupported Content-Type '#{content_type}', expected one of: #{@allowed_content_types.sort.join(', ')}"
          })
          next
        end

        data = req.body.to_s.dup
        extra = { 'HTTP_REQUEST_TARGET_NAME' => target_name, 'HTTP_REQUEST_PACKET_NAME' => packet_name }
        extra['HTTP_HEADERS'] = req.header.transform_values(&:first) if req.header
        extra['HTTP_QUERIES'] = req.query unless req.query.empty?

        # 5. Enqueue with backpressure
        begin
          @request_queue.push([data, extra], true)  # non_block: raises ThreadError when full
        rescue ThreadError
          res.status = 503
          res['Content-Type'] = 'application/json'
          res['Retry-After'] = '5'
          res.body = JSON.generate({ status: 'error', message: 'Server queue full — retry later' })
          next
        end

        res.status = 200
        OK_HEADERS.each { |k, v| res[k] = v }
        res.body = OK_BODY
      end

      @server_thread = Thread.new { @server.start }
      super()
    end

    def connected?
      !@server.nil?
    end

    def disconnect
      @server&.shutdown
      @server_thread&.join
      @server = nil
      @server_thread = nil
      super()
      # Drain first so the sentinel always fits in a bounded queue.
      @request_queue.clear
      @request_queue << nil  # unblock read_interface
    end

    def read_interface
      item = @request_queue.pop
      return nil, nil if item.nil?
      data, extra = item
      read_interface_base(data, extra)
      [data, extra]
    end

    def write_interface(_data, _extra = nil)
      raise 'Commands cannot be sent to HttpJsonServerInterface'
    end

    def convert_packet_to_data(_packet)
      raise 'Commands cannot be sent to HttpJsonServerInterface'
    end

    def convert_data_to_packet(data, extra = nil)
      packet = Packet.new(nil, nil, :BIG_ENDIAN, nil, data.to_s)
      packet.accessor = JsonAccessor.new(packet)
      if extra
        target_name = extra.delete('HTTP_REQUEST_TARGET_NAME')
        packet_name = extra.delete('HTTP_REQUEST_PACKET_NAME')
        if target_name && packet_name
          packet.target_name = target_name.upcase
          packet.packet_name = packet_name.upcase
        end
        packet.extra = extra
      end
      packet
    end

    def details
      result = super()
      result['listen_address'] = @listen_address
      result['port'] = @port
      result['request_queue_length'] = @request_queue.length
      result
    end
  end
end

File: openc3-cosmos-httpjsondemo/targets/HTTPJSONDEMO/lib/http_json_server_interface.py
# Copyright 2026 OpenC3, Inc.
# All Rights Reserved.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE.md for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

import json
import queue
import traceback
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from threading import Thread
from urllib.parse import parse_qsl

from openc3.interfaces.interface import Interface
from openc3.packets.packet import Packet
from openc3.system.system import System
from openc3.utilities.logger import Logger

_OK_BODY = b'{"status":"ok"}'
_OK_HEADERS = [
    ("Content-Type", "application/json"),
    ("Content-Length", str(len(_OK_BODY))),
    ("Connection", "close"),
]


def _error_body(message):
    body = json.dumps({"status": "error", "message": message}).encode("utf-8")
    headers = [
        ("Content-Type", "application/json"),
        ("Content-Length", str(len(body))),
        ("Connection", "close"),
    ]
    return body, headers


class _Handler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):  # suppress default stdout access log
        pass

    def _send(self, status, body, headers, extra_headers=None):
        self.send_response(status)
        if extra_headers:
            for k, v in extra_headers:
                self.send_header(k, v)
        for k, v in headers:
            self.send_header(k, v)
        self.end_headers()
        self.wfile.write(body)

    def do_POST(self):  # noqa: N802
        # 1. Authentication
        api_key = getattr(self.server, "api_key", None)
        if api_key is not None and self.headers.get("x-api-key") != api_key:
            body, headers = _error_body("Unauthorized")
            self._send(401, body, headers)
            return

        # 2. Path — expect /targetname/packetname
        parts = self.path.split("?")[0].strip("/").split("/")
        if len(parts) != 2 or not all(parts):
            body, headers = _error_body(f"Path must be /target/packet, got: {self.path}")
            self._send(400, body, headers)
            return

        target_name = parts[0].upper()
        packet_name = parts[1].upper()

        # 3. Known packet validation
        valid_packets = getattr(self.server, "valid_packets", None)
        if valid_packets is not None and (target_name, packet_name) not in valid_packets:
            body, headers = _error_body(f"Unknown packet: {target_name}/{packet_name}")
            self._send(404, body, headers)
            return

        # 4. Content-Type enforcement
        allowed = getattr(self.server, "allowed_content_types", {"application/json"})
        content_type = self.headers.get("content-type", "").split(";")[0].strip().lower()
        if content_type not in allowed:
            body, headers = _error_body(
                f"Unsupported Content-Type '{content_type}', expected one of: {', '.join(sorted(allowed))}"
            )
            self._send(415, body, headers)
            return

        # 5. Content-Length
        try:
            length = int(self.headers.get("content-length", 0))
        except ValueError:
            body, headers = _error_body("Invalid Content-Length header")
            self._send(400, body, headers)
            return
        if length < 0:
            body, headers = _error_body("Negative Content-Length")
            self._send(400, body, headers)
            return
        data = self.rfile.read(length) if length else b""

        # 6. Build extra metadata
        extra = {
            "HTTP_HEADERS": {k.lower(): v for k, v in self.headers.items()},
            "HTTP_REQUEST_TARGET_NAME": target_name,
            "HTTP_REQUEST_PACKET_NAME": packet_name,
        }
        query_string = self.path.partition("?")[2]
        if query_string:
            queries = dict(parse_qsl(query_string))
            if queries:
                extra["HTTP_QUERIES"] = queries

        # 7. Enqueue with backpressure
        try:
            self.server.request_queue.put_nowait((data, extra))
        except queue.Full:
            body, headers = _error_body("Server queue full — retry later")
            self._send(503, body, headers, extra_headers=[("Retry-After", "5")])
            return

        self._send(200, _OK_BODY, _OK_HEADERS)


class HttpJsonServerInterface(Interface):
    """HTTP server interface that receives inbound JSON POSTs as telemetry.

    Path format: POST /targetname/packetname
    The path segments are uppercased and used to identify the telemetry packet.
    Item names in the packet definition must match the incoming JSON key names.
    Telemetry definitions for packets received by this interface must declare
    ACCESSOR JsonAccessor so that all COSMOS microservices parse the buffer correctly.

    Supported options:
      LISTEN_ADDRESS <ip>  — bind address (default 0.0.0.0)
    """

    def __init__(self, port=80):
        super().__init__()
        self.listen_address = "0.0.0.0"
        self.port = int(port)
        self.server = None
        self.api_key = None
        self.max_queue_depth = 1000
        self.allowed_content_types = {"application/json"}
        self._request_queue = queue.Queue()

    def set_option(self, option_name, option_values):
        super().set_option(option_name, option_values)
        match option_name.upper():
            case "LISTEN_ADDRESS":
                self.listen_address = option_values[0]
            case "API_KEY":
                self.api_key = option_values[0]
            case "MAX_QUEUE_DEPTH":
                self.max_queue_depth = int(option_values[0])
            case "ALLOW_CONTENT_TYPE":
                self.allowed_content_types.add(option_values[0].lower())

    def connection_string(self):
        return f"listening on {self.listen_address}:{self.port}"

    def connect(self):
        self._request_queue = queue.Queue(maxsize=self.max_queue_depth)

        self.server = ThreadingHTTPServer((self.listen_address, self.port), _Handler)
        self.server.request_queue = self._request_queue
        self.server.api_key = self.api_key
        self.server.allowed_content_types = self.allowed_content_types

        # Build the valid-packet lookup so the handler can reject unknown paths before 200 OK.
        valid_packets = set()
        for target_name in self.target_names:
            try:
                for packet_name in System.telemetry.packets(target_name):
                    valid_packets.add((target_name.upper(), packet_name.upper()))
            except Exception:
                Logger.error(
                    f"HttpJsonServerInterface: failed to enumerate packets for {target_name}\n{traceback.format_exc()}"
                )
        self.server.valid_packets = valid_packets

        self._server_thread = Thread(target=self.server.serve_forever, daemon=True)
        self._server_thread.start()
        super().connect()

    def connected(self):
        return self.server is not None

    def disconnect(self):
        if self.server:
            self.server.shutdown()
            self._server_thread.join()
        self.server = None
        super().disconnect()
        # Drain first so the sentinel always fits in a bounded queue.
        while True:
            try:
                self._request_queue.get_nowait()
            except queue.Empty:
                break
        self._request_queue.put((None, None))  # unblock read_interface

    def convert_packet_to_data(self, packet):
        raise RuntimeError("Commands cannot be sent to HttpJsonServerInterface")

    def write_interface(self, data, extra=None):
        raise RuntimeError("Commands cannot be sent to HttpJsonServerInterface")

    def read_interface(self):
        data, extra = self._request_queue.get(block=True)
        if data is None:
            return None, None
        self.read_interface_base(data, extra)
        return data, extra

    def convert_data_to_packet(self, data, extra=None):
        packet = Packet(None, None, "BIG_ENDIAN", None, data)
        if extra:
            target_name = extra.pop("HTTP_REQUEST_TARGET_NAME", None)
            packet_name = extra.pop("HTTP_REQUEST_PACKET_NAME", None)
            if target_name and packet_name:
                packet.target_name = target_name
                packet.packet_name = packet_name
            packet.extra = extra
        return packet

    def details(self):
        result = super().details()
        result["listen_address"] = self.listen_address
        result["port"] = self.port
        result["request_queue_length"] = self._request_queue.qsize()
        return result
